common/usermode.c

```diff
@@ -177,67 +89,63 @@ static void init_syscall(void) {
 void init_usermode(percpu_t *percpu) {
     vmap_user_4k(&cr3, virt_to_mfn(&cr3), L1_PROT);
     vmap_user_4k(&enter_usermode, virt_to_mfn(&enter_usermode), L1_PROT);
-    vmap_user_4k(&syscall_handler, virt_to_mfn(&syscall_handler), L1_PROT);
-
+    vmap_user_4k(&syscall_handler_entry, virt_to_mfn(&syscall_handler_entry), L1_PROT);
     init_syscall();
 }
 
-static inline void __user_text sys_exit(unsigned long exit_code) {
-    asm volatile("syscall" ::"A"(SYSCALL_EXIT), "D"(exit_code) : STR(_ASM_CX), "r11");
-}
-
-static inline long __user_text sys_printf(const char *fmt, unsigned long arg1,
-                                          unsigned long arg2, unsigned long arg3) {
-    register unsigned long rax asm(STR(_ASM_AX));
+static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, long arg3,
+                                       long arg4, long arg5) {
+    register long return_code asm(STR(_ASM_AX));
+    register long _arg4 asm("r8") = arg4;
+    register long _arg5 asm("r9") = arg5;
 
     /* clang-format off */
     asm volatile(
-        "syscall"
-        : "=A"(rax)
-        : "0"(SYSCALL_PRINTF), "D"(fmt), "S"(arg1), "b"(arg2), "d"(arg3)
+        SAVE_CALLEE_SAVED_REGS()
+        "syscall\n"
+        RESTORE_CALLEE_SAVED_REGS()
+        : "=a"(return_code)
```

Comment:
Ahh, I see. You worked around the register clobbering by saving/restoring in userland. But to get the privilege boundaries right, that should be done in the kernel.

Reply:
Ok, as Johannes also suggested doing this in the kernel, I will change accordingly.